src/units/abilities.cpp

```diff
@@ -1537,29 +1537,30 @@ bool attack_type::special_active_impl(const_attack_ptr self_attack, const_attack
 	temporary_facing self_facing(self, self_loc.get_relative_dir(other_loc));
 	temporary_facing other_facing(other, other_loc.get_relative_dir(self_loc));
 
-	// Filter poison, plague, drain, first strike
-	if (tag_name == "drains" && other && other->get_state("undrainable")) {
-		return false;
-	}
-	if (tag_name == "plague" && other &&
-		(other->get_state("unplagueable") ||
-		 resources::gameboard->map().is_village(other_loc))) {
+	// Filter poison, plague, drain, slow, petrifies
+	// True if "whom" corresponds to "self", false if "whom" is "other"
+	bool whom_is_self = ((whom == AFFECT_SELF) || ((whom == AFFECT_EITHER) && special_affects_self(special, is_attacker)));
+	unit_const_ptr whom_unit = whom_is_self ? other : self;
+	map_location whom_loc = whom_is_self ? other_loc : self_loc;
```

Comment:
Sorry, I was confused when suggesting names. `whom_unit` would make sense for the unit that is `whom`, but the unit here is the opponent of `whom`. How about `whom_is_self`, `them_unit` and `them_loc`; alternatively `whom_is_self`, `them`, `their_loc`?

So the comment could be
```
// There are only two units involved in this logic, but three pairs of names for them: {self, other},
// {attacker, defender} and {whom, them}.
```

Reply:
@stevecotton names changed, but i not added comment.